lrlex/src/lib/parser.rs

```diff
@@ -71,37 +90,55 @@ impl<StorageT: TryFrom<usize>> LexParser<StorageT> {
     }
 
     fn parse(&mut self) -> LexBuildResult<usize> {
-        let mut i = self.parse_declarations(0).map_err(|e| vec![e])?;
-        i = self.parse_rules(i).map_err(|e| vec![e]).and_then(|i| {
-            if let Some((orig_span, spans)) = self.duplicate_names.iter().next() {
-                let spans = std::iter::once(*orig_span)
-                    .chain(spans.iter().copied())
-                    .collect::<Vec<Span>>();
-                return Err(vec![LexBuildError {
-                    spans,
-                    kind: LexErrorKind::DuplicateName,
-                }]);
-            }
-            Ok(i)
-        })?;
-        // We don't currently support the subroutines part of a specification. One day we might...
-        match self.lookahead_is("%%", i) {
-            Some(j) => {
-                let k = self.parse_ws(j).map_err(|e| vec![e])?;
-                if k == self.src.len() {
-                    Ok(i)
-                } else {
-                    Err(vec![self.mk_error(LexErrorKind::RoutinesNotSupported, i)])
+        let mut errs = Vec::new();
+        let mut result = self.parse_declarations(0, &mut errs);
+        result = self.parse_rules(
+            match result {
+                Ok(i) => i,
+                Err(e) => {
+                    errs.push(e);
+                    return Err(errs);
                 }
+            },
+            &mut errs,
+        );
+
+        // We don't currently support the subroutines part of a specification. One day we might...
+        match result {
+            Err(e) => {
+                errs.push(e);
+                Err(errs)
             }
-            None => {
-                assert_eq!(i, self.src.len());
-                Ok(i)
-            }
+            Ok(i) => match self.lookahead_is("%%", i) {
+                Some(j) => {
+                    let k = self.parse_ws(j).map_err(|e| vec![e])?;
```

Comment:
this `map_err` should `errs.push(..); ...`, will try and fix these issues tomorrow